### PR TITLE
[WIP] Refactor to use pull streams

### PR DESCRIPTION
**Pull request by dignifiedquire**
- [x] Fix tests
- [x] Fix error handling
- [x] Test in libp2p-swarm
- [x] Cleanup state management
- [x] Fix and publish https://github.com/pull-stream/stream-to-pull-stream/issues/9
- [x] Fix and publish https://github.com/pull-stream/pull-handshake/pull/3

### CODE REVIEW

**Comment by dominictarr** [comment]
the two "fix and publish" tasks are done now.

**Comment by dignifiedquire** [comment]
thanks @dominictarr :) just on my flight back from holidays so will hopefully be able to finish this in the next days

**Comment by dominictarr** [comment]
btw, I quite like how you have used run-series to separate the parts of the handshake.
but I would like to suggest completely separating the crypto from the IO.
I did that here https://github.com/auditdrivencrypto/secret-handshake/blob/master/state.js
but in an OO style, which I will change.
better would be to put the crypto in pure functions (pass in state, message, return new state)
because then you can generate test vectors, and test the implementation in every language you port to, without having to test the IO part.

my plan is to refactor `secret-handshake` to work like that, just gotta make the one random thing [ephemeral key generation](https://github.com/auditdrivencrypto/secret-handshake/blob/master/state.js#L33) accept a seed passed in as initial state.

**Comment by dignifiedquire** [comment]
@dominictarr I'm getting close to working :) but I have this nasty issue when I started using `pull-defer` and I'm not sure if it's a bug in my usage or in the library. 

You can run it yourself with `DEBUG=libp2p* npm run test:node`, and here is the log output: https://gist.github.com/dignifiedquire/fd6623ea88f8b8d2e5cbd4f95abcb9d4
Any pointers would be appreciated :)

**Comment by dominictarr** [comment]
the problem is here:
https://github.com/libp2p/js-libp2p-secio/blob/pull/src/handshake/finish.js#L20-L29

you are trying to read to the same stream in from two different places. that isn't allowed.
with pull streams each stream should be attached to exactly one stream.
(like plumming, if you want to attach two pipes to one, you need a special T or Y connector)

you need to take the stream, and put the mac into the start, and intercept it coming out.
This is part of the "body" encryption, but really it's still a part of the handshake, because you don't truely know if the handshake has succeed until you get the right value out.

``` js
var shake = Handshake()

pull(
   stream,
   etm.createUnboxStream(proto.local.cipher, proto.local.mac),
   shake,
   etm.createBoxStream(proto.remote.cipher, proto.remote.mac),
   stream
)
shake.write(state.proposal.in.rand)
shake.read(state.proposal.in.rand.length, function (err, buf) {
  if(deepEqual(buf, state.proposal.in.rand)) //SUCCESS
    state.secure.resolve(shake.rest())
  else
     state.secure.resolve(pull.error(new Error('fail')))
})
```
it seems silly to use two handshake phases, but since it sends a fixed value to confirm the session, part of the handshake is in the body, so that is kinda just how this protocol works.

**Comment by dignifiedquire** [comment]
@dominictarr thank you so much! I finally have passing tests again 😂

**Comment by dignifiedquire** [comment]
Depends on 
- https://github.com/dignifiedquire/pull-length-prefixed/pull/3
- https://github.com/multiformats/js-multistream/pull/19

**Comment by diasdavid** [comment]
I believe you have successfully tested this in libp2p-swarm, right?

**Comment by dignifiedquire** [comment]
Yes the secio branch in libp2p-swarm adds tests using secio which passed with the current version of this branch

**Comment by dignifiedquire** [comment]
This probably still needs some cleanup, need to go through it once more before merging

File: package.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "name": "libp2p-secio",
   "version": "0.3.1",
   "description": "Secio implementation in JavaScript",
-  "main": "lib/index.js",
+  "main": "src/index.js",
```
Comment (diasdavid) [inline package.json]: - [ ] change before merge

File: src/handshake/finish.js [hunk 2]
```diff
@@ -1,35 +1,51 @@
...
+    try {
+      crypto.verifyNonce(state, nonceBack)
+    } catch (err) {
+      return fail(err)
     }
 
-    log('3. finish - finish', nonceOut.toString('hex'), nonceOut2.toString('hex'))
+    log('3. finish - finish')
 
     // Awesome that's all folks.
```
Comment (diasdavid) [inline src/handshake/finish.js]: 👌🏽

**Comment by diasdavid** [comment]
Last things to do before merge:

- [ ] point to lib
- [ ] rebase commits to follow js-commit guidelines
- [ ] add documentation about it using pull-streams
- [ ] complete the readme (API + Flow of the handshake (even it just a drawing))

**Comment by diasdavid** [comment]
Extra, required for js and go interop https://github.com/libp2p/js-libp2p-secio/issues/4#issuecomment-244619965

**Comment by dignifiedquire** [comment]
> Extra, required for js and go interop #4 (comment)

This should be simple with this: https://github.com/dignifiedquire/pull-length-prefixed/pull/5

**Comment by dignifiedquire** [comment]
- Example is working
- Tests are working except one, but not sure where the failure is coming from:

```
libp2p-ipfs crash that node, make sure nodeA continues going steady:
     Uncaught AssertionError: expected [ Array(4) ] to have a length of 3 but got 4
      at Timeout.check [as _onTimeout] (test/libp2p.spec.js:413:42)
```

**Pull request merged**